File: functions/src/features/quests/types/quest.ts
import {QuestTypeEnum} from "./quest-type-enum";
import {Answer} from "./answer";
import {QuestSubTypeEnum} from "./quest-sub-type-enum";

export type Quest = {
  id: string;
  description: { [key: string]: string };
  points: number;
  validityStart: Date;
  validityEnd: Date;
  executionTime: number;
  type: QuestTypeEnum;
  subType: QuestSubTypeEnum | null;

  // Actor & community quest section
  queueTime: number | null;
  actor: string | null;
  maxQueue: number | null;
  groupSize: number | null;
  requestAccepted: boolean | null;
  promptFunction: string | null;

  // Quiz quest section
  parentQuests: string[] | null;
  qrCode: string | null;
  question: { [key: string]: string } | null;
  answers: Answer[] | null;

  // Social quest section
  verificationFunction: string | null;
}

File: functions/features/quests/types/quest.py
from typing import List, Optional
from pydantic import BaseModel

from features.quests.types.quest_type_enum import QuestTypeEnum
from features.quests.types.quest_sub_type_enum import QuestSubTypeEnum
from features.quests.types.answer import Answer
from features.quests.types.localized_field import LocalizedField
from features.quests.prompt_functions import PromptFunctions
from datetime import datetime


class Quest(BaseModel):
    id: str
    title: Optional[LocalizedField] = None
    description: LocalizedField
    points: List[int]
    validityStart: datetime
    validityEnd: datetime
    executionTime: int
    type: QuestTypeEnum
    subType: Optional[QuestSubTypeEnum] = None

    # Actor & community quest section
    queueTime: Optional[int] = None
    actor: Optional[str] = None
    maxQueue: Optional[int] = None
    groupSize: Optional[int] = None
    requestAccepted: Optional[bool] = None
    promptFunction: Optional[PromptFunctions] = None

    # Quiz quest section
    parentQuests: Optional[List[str]] = None
    qrCode: Optional[str] = None
    question: Optional[LocalizedField] = None
    answers: Optional[List[Answer]] = None

    # Social quest section
    verificationFunction: Optional[str] = None
